hnn_core/externals/bayesopt.py

```diff
@@ -0,0 +1,145 @@
+
+Brochu, Cora, and de Freitas' tutorial at
+http://haikufactory.com/files/bayopt.pdf
+
+Adopted from http://atpassos.me/post/44900091837/bayesian-optimization
+"""
+
+# Authors: Alexandre Gramfort <alexandre.gramfort@telecom-paristech.fr>
+#          Alexandre Passos <alexandre.tp@gmail.com>
+#          Mainak Jas <mainak.jas@telecom-paristech.fr>
+#          Carolina Fernandez <cxf418@miami.edu>
+
+import warnings
+from sklearn import gaussian_process
+import numpy as np
+
+import scipy.stats as st
+
+
+def expected_improvement(gp, best_f, all_x):
+    """The expected improvement acquisition function. The equation is
+    explained in Eq (3) of the tutorial.
+
+    Parameters
+    ----------
+    gp : instance of GaussianProcessRegressor
+        The GaussianProcessRegressor object.
+    best_f : float
+        Best objective value.
+    all_x : ndarray
+        Randomly distributed samples.
+
+    Returns
+    -------
+    ndarray
+        Object values corresponding to all_x.
+    """
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        # (n_samples, n_features)
+        y, y_std = gp.predict(all_x, return_std=True)
+    Z = (y - best_f) / (y_std + 1e-12)
+    return (y - best_f) * st.norm.cdf(Z) + y_std * st.norm.pdf(Z)
+
+
+def bayes_opt(func, x0, cons, acquisition, maxfun=200,
+              debug=False, random_state=None):
+    """The actual bayesian optimization function.
+
+    Parameters
+    ----------
+    func : func
+        The objective function.
+    x0 : list
+       Initial parameters.
+    cons : list of tuples
+        Parameter constraints in solver-specific format.
+    acquisition : func
+        Acquisiton function we want to use to find query points.
+    maxfun : int, optional
+        Maximum number of function evaluations. The default is 200.
+    debug : bool, optional
+        The default is False.
+    random_state : int, optional
+        Random state of the GaussianProcessRegressor. The default is None.
+
+    Returns
+    -------
+    best_x : list
+        The argument that minimizes f.
+    best_f : float
+        The final objective value.
+    """
+
+    X, y = list(), list()
+
+    # evaluate
+    initial_f = func(x0)
+    if not np.isinf(initial_f):
+        X.append(x0)
+        y.append(initial_f)
+
+    best_x = X[np.argmin(y)]
+    best_f = y[np.argmin(y)]
+    gp = gaussian_process.GaussianProcessRegressor(random_state=random_state)
+
+    if debug:
+        print("iter", -1, "best_x", best_x, best_f)
+
+    for i in range(maxfun):
+
+        # draw samples from distribution
+        all_x = np.random.uniform(low=[idx[0] for idx in cons],
+                                  high=[idx[1] for idx in cons],
+                                  size=(10000, len(cons)))
+
+        # (n_samples, n_features)
+        gp.fit(np.array(X), np.array(y))
+
+        # get new set of params
+        new_x = all_x[acquisition(gp, best_f, all_x).argmin()]  # lowest obj
```

Comment:
I was looking at the implementation of the blog post here: https://gist.github.com/alextp/1402892

and found one difference between the original and ours: it uses annealing from `scipy.optimize`. That is a deprecated function but this is recommended in its place:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.basinhopping.html#scipy.optimize.basinhopping

Reply:
skopt seems to use lbfgs-b as the default optimizer: https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html

You could try with either ... I can see why sampling may give non-optimal results.